lib/src/web/database.dart

```diff
@@ -0,0 +1,282 @@
+              // autocommit state
+              return callback(_ExclusiveTransactionContext(this, writeContext));
+            }),
+        debugContext: 'writeTransaction()',
+        lockTimeout: lockTimeout);
+  }
+
+  @override
+
+  /// Internal writeLock which intercepts transaction context's to verify auto commit is not active
+  Future<T> writeLock<T>(Future<T> Function(SqliteWriteContext tx) callback,
+      {Duration? lockTimeout, String? debugContext}) async {
+    if (_mutex case var mutex?) {
+      return await mutex.lock(() async {
+        final context = _ExclusiveContext(this);
+        try {
+          return await callback(context);
+        } finally {
+          context.markClosed();
+        }
+      });
+    } else {
+      // No custom mutex, coordinate locks through shared worker.
+      await _database.customRequest(CustomDatabaseMessage(
+          CustomDatabaseMessageKind.requestExclusiveLock));
+      final context = _ExclusiveContext(this);
+      try {
+        return await callback(context);
+      } finally {
+        context.markClosed();
+        await _database.customRequest(
+            CustomDatabaseMessage(CustomDatabaseMessageKind.releaseLock));
+      }
+    }
+  }
+}
+
+class _SharedContext implements SqliteReadContext {
+  final WebDatabase _database;
+  bool _contextClosed = false;
+
+  _SharedContext(this._database);
+
+  @override
+  bool get closed => _contextClosed || _database.closed;
+
+  @override
+  Future<T> computeWithDatabase<T>(
+      Future<T> Function(CommonDatabase db) compute) {
+    // Can't be implemented: The database may live on another worker.
+    throw UnimplementedError();
+  }
+
+  @override
+  Future<Row> get(String sql, [List<Object?> parameters = const []]) async {
+    final results = await getAll(sql, parameters);
+    return results.single;
+  }
+
+  @override
+  Future<ResultSet> getAll(String sql,
+      [List<Object?> parameters = const []]) async {
+    return await wrapSqliteException(
+        () => _database._database.select(sql, parameters));
+  }
+
+  @override
+  Future<bool> getAutoCommit() async {
+    return _database.getAutoCommit();
+  }
+
+  @override
+  Future<Row?> getOptional(String sql,
+      [List<Object?> parameters = const []]) async {
+    final results = await getAll(sql, parameters);
+    return results.singleOrNull;
+  }
+
+  void markClosed() {
+    _contextClosed = true;
+  }
+}
+
+class _ExclusiveContext extends _SharedContext implements SqliteWriteContext {
+  _ExclusiveContext(super.database);
+
+  @override
+  Future<ResultSet> execute(String sql,
+      [List<Object?> parameters = const []]) async {
+    return wrapSqliteException(
+        () => _database._database.select(sql, parameters));
+  }
+
+  @override
+  Future<void> executeBatch(
+      String sql, List<List<Object?>> parameterSets) async {
+    return wrapSqliteException(() async {
+      for (final set in parameterSets) {
```

Comment:
this is more for my own understanding but is there a reason we don't run these with something like Promise.all dart equivalent (I think we don't do that in JS either) so just curious

Reply:
I'm not entirely sure if that would be the use case for this. `Future.wait` makes error handling a little more difficult i.e knowing which future or query caused the error.

It would also require us rewriting this to make a list of futures with the parameter sets. I don't really see the advantage in changing it.

Reply:
It lets things run in parallel as opposed to sequentially but if it's not an issue then don't think it's something to worry about